### DISCUSSION TITLE
Notes on required features for a Symbolic Quantum API

### DISCUSSION THREAD

**Discussion by sritchie** [sicmutils/sicmutils#270] [Ideas]
Let's talk about what features we'd need, and what we'd DO with, a symbolic quantum API.

Here is a great note kicking this off from @CTW in Clojurians Slack:

New thread...
So I’ve finally sat down to start thinking about an implementation of quantum
mechanics using sicmutils. I don’t know how familiar other people are with the
maths (or concepts) but I’ve made a quick (1 page) attempt at trying to nail down the bare essentials of what
we’d need in the toolbox to really get going. I had a look at the docs you pinned
but I’m very new to Clojure, so please bear with me.
Quantum mechanics is big on complex numbers and matrices, but I think
you have that covered. The next big thing are ’operators’ and, from what I can
see, something also already exists for this, but it might need some abstraction
to allow for dropping down to matrix methods? The big one though are ’kets’,
used to represent the state of a system. These are really infinite-dimensional
vector spaces and I’m not sure how best to represent these (and their duals).
For lower-level calculations, they can be represented as ’up’ states but, ideally,
we should keep the higer level of abstraction as long as possible.

For the very basics I've listed some maths examples in a pdf. Could someone have a look and tell me what already exists to start with? (edited) 

[slackmessage.pdf](https://github.com/sicmutils/sicmutils/files/5968456/slackmessage.pdf)

"Is the best way to model something like the ket, to start with a map of labels and then define protocols for operating on it or can I 'extend' an existing type?"

And a Zulip followup:

![image](https://user-images.githubusercontent.com/69635/107698635-0f146b00-6c72-11eb-9dbc-d0b5b8e7a530.png)

**Comment by sritchie** [comment]
It seems like the place to start is implementing `abstract-up` and `abstract-down` from [scmutils](https://github.com/Tipoca/scmutils/blob/master/src/kernel/structs.scm#L537), and maybe getting them to render as ket and bra.

We also need:

- kronecker product
- proper symbolic operations on the abstract structures for inner, outer product, transpose, conjugate etc

And then of course as you say, maybe there is some different thing, or a differently-argumented thing, that lets us talk about infinite dimensional states.

**Reply by MagusMachinae** [reply]
Hi Sam, sorry it took me this long to get back to you. New year's been a little hectic for me and I've been having trouble devoting as much time and focus to this as I'd like.

I wasn't quite sure how to get started on the symbolic operations themselves, so I decided to start with implementing functions that return a state vector given a state number (any base) using the structure object like so:

<img width="257" alt="Discord_KaayABXzti" src="https://user-images.githubusercontent.com/47104486/107784580-8960f000-6d43-11eb-98a7-b2abc494f305.png">

IIRC one thing that gave me pause with this implementation, as well as trying to simply use `up` or `down` to represent the vector was that the actual data structure isn't quite homoiconic ie. a column vector still looks like `[a b c d]` instead of something columnar. I think I first tried using the `Matrix` protocol to get that representation before trying this one.

I was a bit unsure of my own knowledge at times, so I'm trying to do a reverse SICM-utils by reading a text book and implementing what I need to as I work through it as a general guide. I've got a copy of Rieffel and Polak's Quantum Computing: A General Introduction and I've been reading QCVC to help keep me on the rails.

The Kronecker product seems pretty straightforward to implement, mapping each of the elements of one nested vector on to another in Clojure-speak. I can have a go playing with that over the next few days and let you know how it goes.

**Reply by light-matters** [reply]
> It seems like the place to start is implementing `abstract-up` and `abstract-down` from [scmutils](https://github.com/Tipoca/scmutils/blob/master/src/kernel/structs.scm#L537), and maybe getting them to render as ket and bra.
> 
> We also need:
> 
>     * kronecker product
> 
>     * proper symbolic operations on the abstract structures for inner, outer product, transpose, conjugate etc
> 
> 
> And then of course as you say, maybe there is some different thing, or a differently-argumented thing, that lets us talk about infinite dimensional states.

Do you know of any examples of how abstract up and down were used by Sussman etc.?

**Comment by light-matters** [comment]
Okay, so to try and get started with this again I've made a public repo: https://gitlab.com/LightMatters/peek.

For the sake of others (this applies below as well), I would just like to clarify that I'm not a computer scientist or a proper programmer. This means that if you think I'm using gitlab/clojure etc. in a crazy way then you're probably right; please let me know!

As for what I'd DO with a symbolic quantum API, hopefully lots of things. I'm a researcher in optical and atomic physics and so I'd like to be able to make symbolic calculations, solve systems of equations and also integrate the associated Clojure expressions with my notes and reports. Another thing that I'm very interested in is the visualisation of results. Combining all of these things together, I would also like to potentially make an interactive course for learning quantum physics, assuming there is ever enough time left in the day :).

**Reply by light-matters** [reply]
Could @sritchie (or someone) maybe explain how easy it would be to extend the structure type? I know Clojure doesn't really do inheritance in the Java sense but is there something vaguely similar using protocols and the 'derive' function?

**Reply by sritchie** [reply]
I can help with this later tonight MST, with lots of pointers to how to do this!

**Reply by MagusMachinae** [reply]
Would it be alright if I sat in for that?

**Reply by light-matters** [reply]
That'd be great but I'm struggling to find a time 'tonight' (for you) that I won't be sleeping :). I'm in GMT+1 and have someone using the guest bedroom (where my desktop is) I can be free in 11 hours but that's a bit late for you. I could maybe do an hour before at a stretch but another day is probably better for me. Don't let me stop the two of you meeting though!

**Reply by MagusMachinae** [reply]
That's kinda convenient actually. I'm in GMT-0, but I'll be working late tonight anyway. I can take notes and then we can go over them at a less nocturnal hour for you.

**Reply by sritchie** [reply]
Hey all! A call is a better idea than the word-dump I was planning, so here is a suggestion. How about we move to tomorrow AM my time, and we can all do this together? If it's useful we could record it as a short tour for later info / posterity, if I can figure out how to do that.

"tonight" today (Sunday) would mean post 8pm my time, so I can get my 2 year old to bed, and that's probably a bit wild.

How about anywhere from 6-9am MST tomorrow AM? Let me know what works for you both! (We can also chat in Slack or Zulip, I'm around in both.) This is a good exercise in moving from Github to more sync communication, thanks for bearing with me :) Either way I"m at your service to make this happen!

**Reply by light-matters** [reply]
That would be great for me. I'm happy to call it 9am MST for clarity but I can make any of the 6-9 option work. 

Exciting times! Thanks very much for offering this, Sam.

P.S. I will also be keeping an eye on Zulip from now.

**Reply by MagusMachinae** [reply]
9AM should be fine for me. I have to take a call at 8 but I don't see it lasting that long.

**Reply by sritchie** [reply]
Thanks all - do you both mind sending me a note at sritchie09@gmail.com? I'll send a zoom invite from there. Otherwise I'll post a link here in the AM when I'm up.

**Reply by sritchie** [reply]
Sending now!

**Reply by light-matters** [reply]
Here's a useful resource: The mathematics of quantum mechanics, Martin Laforest. It's a downloadable textbook that might come in handy. It's surprisingly hard to google baseless linear algebra references but this is a start.

**Reply by sritchie** [reply]
@MagusMachinae , @light-matters , I've got a Discord server I host for the newsletter I created a while back; there are some folks there interested in QM + SICM, so let's use that as a development / screenshare home. Here's the link: https://discord.gg/pJcCrHp6

**Comment by sritchie** [comment]
Here are some notes, @MagusMachinae and @light-matters.

First, @alexgian just alerted us to @JonyEpsilon's work, in particular [A term-rewriting system for computer quantum algebra](https://arxiv.org/abs/0809.4466) back from 2008. What a beacon! This is quite close, if not identical, to what we want to build into SICMUtils.

How to do it? Here was a sketch of the path I suggested:

Step one is to come up with a first twitch of life we'd like to see out of this system. That means coming up with the data structures and the operations between them. @light-matters notes: "I think there are really two forms of abstract in regards to the tensors. Coordinate (basis) independence and 'infiniteness'. I think basisless operations are the clearest entry point so I would start with that."

So we want a `bra` and `ket` constructor, probably. Eventually we'll build out nice types using `deftype` that can implement all of the protocols in SICMUtils, but for now it doesn't matter. We are doing Wishful Thinking Programming. Start with the menu of constructor functions we want, plus a few operations between them. Something like

```clojure
(defn bra [& components] ...)
(defn ket [& components] ...)
```

maybe backed by a map like `{:type :bra :components [0 1 1]}` or something for now.

Then, write down the menu of operations between these types.

Finally, write a "print" function that will kick out the proper Dirac-notation rendering of the thing.

That should be enough for a tiny pull request that we can all stare at together! I was going to write a longer post about how to get a type installed into the sicmutils generic system... but I don't think that is necessary at this point. Go ahead and gleefully duplicate with new, non-generic functions called "inner-product" etc, and then once we get something working we can wire it in.

Thoughts?

**Reply by light-matters** [reply]
I don't have a lot of time to look at this at the moment but just wanted to say: great! The paper particularly, looks very useful indeed. I'll have to print it out and have a nosy over the next few days.

I actually started doing something like what you suggest in 'state.clj' but I've spent almost no time on it. Just a couple of functions about creating kets and bras from components and labels and from one another.

Exciting times. The paper should give us something concrete to build on and that's always good for motivation.

**Comment by sritchie** [comment]
@light-matters @MagusMachinae here are some other folks on to the same stuff:

### Quantomatic 

- https://github.com/Quantomatic
  - many referenced papers: https://quantomatic.github.io/papers.html
  - a summary of their software: http://swmath.org/software/10107

### Sympy support:

- overview: https://digitalcommons.calpoly.edu/cgi/viewcontent.cgi?article=1039&context=physsp
- module docs: https://docs.sympy.org/latest/modules/physics/quantum/index.html

example from the overview:

![image](https://user-images.githubusercontent.com/69635/109550553-af4cfb00-7a8c-11eb-9342-d59a5fcf3531.png)

### Mathematica

SQCS: http://www.swmath.org/software/10188

**Comment by light-matters** [comment]
Just adding Sam's suggestion: [Visual wave functions](https://ridiculousfish.com/wavefiz).

And a paper on a Haskell attempt: [SIQM](https://dl.acm.org/doi/10.1145/871895.871901).
